python/graphstorm/gpartition/post_hard_negative.py

```diff
@@ -0,0 +1,109 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+"""
+
+import json
+import os
+
+import torch as th
+from dgl.data.utils import load_tensors, save_tensors
+from graphstorm.model.utils import load_dist_nid_map
+
+
+def load_hard_negative_config(gsprocessing_config):
+    """Load GSProcessing Config to extract hard negative config
+
+    Parameters
+    ----------------
+    gsprocessing_config: str
+        Path to the gsprocessing config.
+    """
+    with open(gsprocessing_config, 'r', encoding='utf-8') as file:
+        config = json.load(file)
+
+    # Hard Negative only supports link prediction
+    edges_config = config['graph']['edges']
+    mapping_edge_list = []
+    for single_edge_config in edges_config:
+        if "features" not in single_edge_config:
+            continue
+        feature_dict = single_edge_config["features"]
+        for single_feature in feature_dict:
+            if single_feature["transformation"]["name"] \
+                    == "edge_dst_hard_negative":
+                edge_type = ":".join([single_edge_config["source"]["type"],
+                                     single_edge_config["relation"]["type"],
+                                     single_edge_config["dest"]["type"]])
+                hard_neg_feat_name = single_feature['name']
+                mapping_edge_list.append({"dst_node_type": single_edge_config["dest"]["type"],
+                                          "edge_type": edge_type,
+                                          "hard_neg_feat_name": hard_neg_feat_name})
+    return mapping_edge_list
+
+
+def shuffle_hard_negative_nids(gsprocessing_config, num_parts, output_path):
+    """Shuffle hard negative edge feature ids with int-to-int node id mapping.
+    The function here align with the shuffle_hard_nids in graphstorm.gconstruct.utils.
+    Create an additional function to handle the id mappings under distributed setting.
+
+    Parameters
+    ----------------
+    gsprocessing_config: str
+        Path to the gsprocessing config.
+    num_parts: int
+        Number of parts.
+    output_path: str
+        Path to the output DGL graph.
+    """
+    shuffled_edge_config = load_hard_negative_config(gsprocessing_config)
+
+    node_type_list = []
+    for single_shuffled_edge_config in shuffled_edge_config:
+        node_type = single_shuffled_edge_config["dst_node_type"]
+        node_type_list.append(node_type)
+    node_mapping = load_dist_nid_map(f"{output_path}/dist_graph", node_type_list)
```

Comment:
Why `f"{output_path}/dist_graph"`?

Reply:
For gspartition, the generated dgl graph will be stored under subdirectory ``dist_graph``.